src/main/java/org/testng/internal/Parameters.java

```diff
@@ -233,19 +260,27 @@ private static void checkParameterTypes(String methodName,
       }
     }
 
-    if ( (parameterNames.length ==0) && (totalLength != 0) ) {
+
+    if ((parameterNames.length == 0)) {
       //parameterNames is usually populated via the @Parameters annotation, so we would need to
       //apply our logic only when @Parameters annotation is not involved.
-      String errPrefix = "Cannot inject " + methodAnnotation + " annotated Method [" + methodName + "] with "
-            + Arrays.toString(parameterTypes);
-      if (mapping.containsKey(methodAnnotation)) {
-        errPrefix = "Can inject only one of " + prettyFormat(mapping.get(methodAnnotation)) +
-            " into a " + methodAnnotation + " annotated " + methodName;
+      String annotation = methodAnnotation;
```

Comment:
It looks `annotation` is only used in the `if (invalid)` so this part could be moved there (and we will gain a bit)